### DISCUSSION TITLE
Training/simulation crashing, possibly due to NaN values

### DISCUSSION THREAD

**Discussion by jackzeng-robotics** [isaac-sim/IsaacLab#1291] [Q&A] [answered]
Hi developers,

I am working on a setup where multiple drones carry an object together with strings. The strings are modeled as 7 thin links with ball joints between them. The ball joints are modeled as 3 continuous joints in x, y and z direction. It looks like this:

![image](https://github.com/user-attachments/assets/1fde10cc-3961-4fa1-b92b-c75a3cbace64)

The links have a very low mass and inertia and before I thought this could create some stability issues. What I observed: given big forces and torques, the setup makes very wild movements and then the setup disappears from the simulation. Probably due to some NaN values. This also caused my training to crash. My hypothesis before was that the cable links have very low mass and inertia causing them to fly away when under high loads, for example when compressed. So, I added numerous termination terms that stopped this from happening, for example limiting the angle between the drone and the links, setting terminations on high velocity values, angular rates etc. 

This pretty much solved the issue for a long time until I implemented a low level controller for the drones to go with it. The behaviour of the low level controller seems fine, similar to what we get in real life. Without the RL policy is does not exhibit any strange behaviour, moreover the actions (forces on the drone rotors) are clamped as well. Now, the error has come back and I get this error when training with SKRL:

[Error] [omni.physx.plugin] PhysX error: The application needs to increase PxGpuDynamicsMemoryConfig::foundLostAggregatePairsCapacity to -1423966208, otherwise, the simulation will miss interactions, FILE /builds/omniverse/physics/physx/source/gpubroadphase/src/PxgAABBManager.cpp, LINE 1269

And this when using RSL-RL:
  File "/home/isaac-sim/.local/share/ov/pkg/isaac-sim-4.2.0/exts/omni.isaac.ml_archive/pip_prebundle/torch/distributions/normal.py", line 71, in sample
    return torch.normal(self.loc.expand(shape), self.scale.expand(shape))
RuntimeError: normal expects all elements of std >= 0.0

Is there any way to figure out where the crash comes from? Like finding the last state of the drones in some log somewhere? Or do you have any idea on what it could be?

Thanks in advance!

Jack

**Comment by StrainFlow** [comment]
Have you reported this in the Issues section? That would be the better place to post a crash bug like this. 

In order for them to solve this they'll need to be able to reproduce the issue, so the more info you have ready to provide the better.

**Comment by DavidTer756** [comment]
Hello Jack,

I hope you're doing well. I came across your work on a similar system in Isaac Lab and noticed that you had encountered an issue quite like the one I’m currently facing.

In my setup, I’ve implemented four drones carrying a cargo, with cables modeled as 13 linked segments connected by joints. During the simulation, the joints sometimes reach extremely high velocities, which causes instability and eventually results in NaN values.

I was wondering if you might be willing to share how you approached or resolved this issue. I would really appreciate any insights or guidance you could offer.

Thank you in advance for your time and help!

**Reply by jackzeng-robotics** [reply, marked as answer]
Hi David, 

I do not have a clean answer for the problem unfortunately. It is however not an uncommon problem, especially in contact based rigid body simulations. I got pretty robust performance when I:
- Decreased the simulation timestep enough (I ran it at 200-300 Hz)
- Terminated the episode before aggressive situations could arise (terminate based on cable angle, payload angle, height, drone angle etc.. before the bodies start to collide with eachother)
- When the setup for my RL policies became better it also happened less since these dangerous were less common

Hope this helps!
Kind regards,
Jack

**Reply by DavidTer756** [reply]
Hi Jack,

I implemented some episode terminations before aggressive situations, and it helped a lot. Thank you very much for your help, I really appreciate it.

Best regards,
David

**Reply by jackzeng-robotics** [reply]
Glad to hear it helped!

**Comment by cyicz123** [comment]
Hi Jack and David,
I ran into exactly the same RuntimeError: normal expects all elements of std >= 0.0 crash during RSL-RL training on a GO2-PIPER quadruped arm manipulating a door handle. After a deep debugging session, the root cause turned out to be the same class of problem you both described: stiff contact on very small-mass bodies → PhysX penetration-recovery force divergence → joint velocity explosion → NaN propagation.
## The causal chain
In my case, the gripper fingers are prismatic links with mass ~0.0225 kg. During early RL exploration, the arm pushes the finger into the door handle at high speed. With the default solver settings (4 position iterations, 0 velocity iterations) and auto contact offsets (5-10 mm on the small convex-decomposition hulls), the finger tunnels into the handle before contact is even generated. Then the penetration recovery kicks in with forces of 5×10⁴ – 7×10⁶ N on a 22-gram body — joint velocity immediately goes to 1×10⁶ rad/s, and the next Normal(std) call gets a negative or NaN std from the exploded value function.
## What I tried (and what didn't work alone)
1. Adding armature to the finger joints — joint-level damping doesn't stop the contact force blow-up at the PhysX level. Runtime logs showed link7/link8 still spiking to 1×10⁶+ N even with armature up to 0.1.
2. Early termination on velocity/angle thresholds — helps reduce crash frequency but doesn't fix the root physics instability.
## What actually fixed it
Two changes to ArticulationCfg, both targeting the penetration root rather than downstream symptoms:
1. Manipulation-grade solver iteration counts (instead of legged-locomotion defaults):
```
solver_position_iteration_count=16,  # was 4
solver_velocity_iteration_count=1,   # was 0 — this is the critical one
```
With 0 velocity iterations, the penetration-recovery velocity is left completely uncorrected, so even a single substep of deep penetration produces unbounded forces. Isaac Lab's Franka/UR10 manipulation tasks all use 16/1 for this reason.

2. Explicit contact offsets to widen the speculative-contact band and prevent tunneling:
```
collision_props=sim_utils.CollisionPropertiesCfg(
    collision_enabled=True,
    contact_offset=0.02,   # catch approach earlier, before penetration
    rest_offset=0.005,     # finger rests ~5 mm outside handle surface
)
```
The door handle is a thin convexHull bar (~2.9 cm thick at our scene scale). PhysX-auto offsets (~5-10 mm) are too small for the thin geometry; the finger tunnels through before contact is generated.
The combination of 16/1 iterations + explicit offsets dropped the peak contact force on the finger links from 7×10⁶ N to below 500 N, and the NaN crash completely disappeared.

## Summary
If your cable/drone system has thin links with very low mass, and you're seeing NaN during training with the default solver settings, try:
- Set solver_velocity_iteration_count=1 at minimum
- Explicitly set contact_offset and rest_offset on the collision bodies
- Consider solver_position_iteration_count=16 for stiff contact scenarios
This is fundamentally a PhysX stiff-contact-on-tiny-mass problem, not an RL bug. The std >= 0 error is just the messenger.

Hope this helps someone debugging the same crash in the future!

**Reply by jackzeng-robotics** [reply]
Hi thanks for investigating this more and letting us know! I will definitely keep this in mind for the future
